app/views/admin/loans/content/_questionnaire_group.html.slim

```diff
@@ -1,34 +1,53 @@
   div
-    - group.children.each_with_index do |field, idx|
-
-      .question data-level=field.level data-type=field.data_type data-id=field.id
-
-        - value = value_set.custom_value(field.id)
+    - optional_encountered = false
+    - optionals_exist = group.children.any? { |i| !i.required_for?(@loan) }
+
+    - group.children.sort_by_required(@loan).each_with_index do |field, idx|
+    
+      - if field.data_type == 'group' && !field.first_child? && !value_set.unanswered?(field)
+        / Close and reopen fieldset with new group/page
+        - if optional_encountered
+          </fieldset>
+          - reopen_fieldset = true
+          
+        / Page header for print view
+        .header
+          = @loan.division.name
+          br
+          = t("loan.print_header_title")
+          br
+          = @loan.organization.name
+          br
+          = I18n.l(Date.today, format: :long)
+
+      - optional = !field.required_for?(@loan)
+      - if reopen_fieldset || (!optional_encountered && optionals_exist && optional && !value_set.unanswered?(field))
+        - optional_encountered = true; reopen_fieldset = false
+        <fieldset class="optional_questions"><legend>#{t 'questionnaires.optional_questions'}</legend>
```

Comment:
This stuff is looking a big gnarly. Could you perhaps do the grouping in the model or controller so that the view doesn't have to do this `optional_encountered` thing?